File: python/flow.py
from map import * ## "Parallel Map Algebra" package
import sys

setupDevices("",0x04,"") # 0x04 == GPU

## Arguments

assert len(sys.argv) > 3
in1 = sys.argv[1]
in2 = sys.argv[2]
out = sys.argv[3]

def flowAccu(water,flow):
	assert( water.numdim() == D2 )
	assert( flow.numdim() == D2 )
	assert( flow.datatype() == U8 )

	acu = zeros_like(water)
	while water:
		ngb = dir2coord(flow)
		acu += water # registers the passing water
		water = spreadSum(water,ngb)

	return acu
# flowAccu

## Computation

water = read(in1)
flow = read(in2)

accu = flowAccu(water,flow)

write(accu,out)

##

acu = zeros_like(water)
delta water:
	ngb = dir2coord(flow)
	acu += water # registers the passing water
	water = spreadSum(water,ngb)

## TOPO op

acu = intrinsic_flowAccu(water,flow)

File: bin/flow.cpp
/**
 * @file    flow.cpp
 * @author  Jesús Carabaño Bravo <jcaraban@abo.fi>
 *
 * Sample 6
 * Spreading operations
 *
 * Calculates the flow accumulation in a DEM for a constant water body of 1xCell
 */

#include "../map.hpp"
using namespace map;
#include <iostream>
#include <string>
#include <cstdio>


int main(int argc, char **argv) {
    Raster dem, dir, pit, water, acu;
    std::string in_file_path, out_file_path;
    int N = 4;

    //// Arguments

    assert(argc > 2);
    in_file_path = std::string(argv[1]);
    out_file_path = std::string(argv[2]);
    if (argc > 3) N = atoi(argv[3]);

    //// Configuration

    setupDevices("",DEV_GPU,"");
    //setupDevices("Intel",DEV_CPU,"");
    //detail::Runtime::getConfig().setNumRanks(1);

    //// Computation
    
    dem = read(in_file_path);
    dir = flowDir(dem);
    water = ones_like(dem,U32) + zeros_like(dem,U32);
    acu = flowAccumulation(water,dir) + zeros_like(dem,U32);
    write(acu, out_file_path);
}
